File: android/app/src/main/java/didgy/dengine/DEngineActivity.java
package didgy.dengine;

import android.app.NativeActivity;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.SurfaceHolder;
import android.view.inputmethod.InputMethodManager;

import java.util.ArrayList;

public class DEngineActivity extends NativeActivity {

    static {
        System.loadLibrary("dengine");
    }

    // Return type is the pointer to the backend-data
    public native double nativeInit(float fontScale);

    public static final int NATIVE_TEXT_JOB_MEMBER_COUNT = 4;
    /*
        This will send multiple text input events to
        the native code at once, the actual text data is in
        the String parameter. The int array is actually a struct-array
        similar to
        struct {
            int oldStart
            int oldCount
            int textOffset
            int textCount
        }
    */
    public native void nativeOnTextInput(
        double backendDataPtr,
        int[] infos,
        String text);
    public static class NativeTextInputJob {
        int start;
        int count;
        CharSequence text;
        NativeTextInputJob(int start, int count, CharSequence text) {
            assert start >= 0;
            this.start = start;
            this.count = count;
            this.text = text;
        }
    }
    // Input text jobs can be submitted in batches, therefore
    // we want signal all of them at once when that happens.
    // This function will turn the data into the linear arrays
    // that the native code expects
    public void nativeOnTextInputWrapper(NativeTextInputJob[] inputJobs) {

        var stringBuilder = new StringBuilder();
        var tempInts = new ArrayList<Integer>();

        for (var item : inputJobs) {
            assert item.text != null;
            assert item.count != 0 || item.text.length() != 0;

            var stringOffset = stringBuilder.length();

            // Push our four members
            tempInts.add(item.start);
            tempInts.add(item.count);
            tempInts.add(stringOffset);
            tempInts.add(item.text.length());
            stringBuilder.append(item.text);
        }
        assert tempInts.size() % NATIVE_TEXT_JOB_MEMBER_COUNT == 0;
        int[] arr = tempInts.stream()
                .mapToInt(i -> i)
                .toArray();
        var outString = stringBuilder.toString();
        nativeOnTextInput(
                mNativeBackendDataPtr,
                arr,
                outString);
    }
    public native void nativeSendEventEndTextInputSession(double backendDataPtr);
    public native void nativeOnNewOrientation(double backendDataPtr, int newOrientation);
    public void nativeOnNewOrientationWrapper(int newOrientation) {
        nativeOnNewOrientation(mNativeBackendDataPtr, newOrientation);
    }
    public native void nativeOnFontScale(double backendDataPtr, int windowId, float newScale);
    public void nativeOnFontScaleWrapper(float newScale) {
        assert mWindowId != INVALID_WINDOW_ID;
        nativeOnFontScale(mNativeBackendDataPtr, mWindowId, newScale);
    }
    public native void nativeOnContentRectChanged(
            double backendDataPtr,
            int posX,
            int posY,
            int width,
            int height);
    public void nativeOnContentRectChangedWrapper(
            int posX,
            int posY,
            int width,
            int height) {
        nativeOnContentRectChanged(
                mNativeBackendDataPtr,
                posX, posY,
                width, height);
    }

    public double mNativeBackendDataPtr = 0;
    public NativeView mNativeContentView = null;
    public Configuration mCurrConfig = null;
    public static final int INVALID_WINDOW_ID = -1;
    public int mWindowId = INVALID_WINDOW_ID;

    int mLastContentX = 0;
    int mLastContentY = 0;
    int mLastContentWidth = 0;
    int mLastContentHeight = 0;

    @Override
    protected void onCreate(Bundle savedState) {

        super.onCreate(savedState);

        mNativeContentView = new NativeView(this);

        setContentView(mNativeContentView);
        mNativeContentView.getViewTreeObserver().addOnGlobalLayoutListener(this);
        mNativeContentView.setFocusable(true);
        mNativeContentView.setFocusableInTouchMode(true);

        var config = new Configuration(getResources().getConfiguration());
        mCurrConfig = config;

        var metrics = getResources().getDisplayMetrics();

        var fontScale = config.fontScale;
        mNativeBackendDataPtr = nativeInit(fontScale);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        assert mCurrConfig != null;

        super.onConfigurationChanged(newConfig);

        int diff = newConfig.diff(mCurrConfig);

        if ((diff & ActivityInfo.CONFIG_FONT_SCALE) != 0) {
            nativeOnFontScaleWrapper(newConfig.fontScale);
        }

        if ((diff & ActivityInfo.CONFIG_ORIENTATION) != 0) {
            nativeOnNewOrientationWrapper(newConfig.orientation);
        }

        mCurrConfig = newConfig;
    }

    @Override
    public void onGlobalLayout() {
        super.onGlobalLayout();

        final int[] newLocation = new int[2];

        mNativeContentView.getLocationInWindow(newLocation);
        int w = mNativeContentView.getWidth();
        int h = mNativeContentView.getHeight();
        if (newLocation[0] != mLastContentX || newLocation[1] != mLastContentY
                || w != mLastContentWidth || h != mLastContentHeight) {
            mLastContentX = newLocation[0];
            mLastContentY = newLocation[1];
            mLastContentWidth = w;
            mLastContentHeight = h;
            nativeOnContentRectChanged(
                    mNativeBackendDataPtr,
                    mLastContentX,
                    mLastContentY,
                    mLastContentWidth,
                    mLastContentHeight);
        }
    }

    @Override
    public void onContentChanged() {
        super.onContentChanged();
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        super.surfaceCreated(holder);
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        super.surfaceChanged(holder, format, width, height);
    }

    // This is called from the native main game thread.
    public void nativeEvent_openSoftInput(String text, int softInputFilter) {
        runOnUiThread(() ->
                this.mNativeContentView.beginInputSession(
                    text,
                    NativeInputFilter.fromNativeEnum(softInputFilter)));
    }

    public void nativeEvent_hideSoftInput() {
        var temp = mNativeContentView;
        runOnUiThread(temp::endInputSession);
    }

    public void nativeEvent_createWindow(int windowId) {
        assert windowId != INVALID_WINDOW_ID;
        assert mWindowId == INVALID_WINDOW_ID;
        mWindowId = windowId;
    }
}

class BeginInputSessionInfo {
    public DEngineActivity mActivity = null;
    public NativeView mView = null;
    public String mText = null;
    public NativeInputFilter inputFilter = null;
    public InputMethodManager mIMM = null;

    public int mSelStart = 0;
    public int mSelCount = 0;

    public int getSelStart() { return mSelStart; }
    public int getSelCount() { return mSelCount; }
    public int getSelEnd() { return mSelStart + mSelCount; }
}

File: android/app/src/main/java/didgy/dengine/DEngineActivity.kt
package didgy.dengine

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.*
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityManager
import androidx.core.view.WindowCompat

class DEngineActivity :
    Activity(),
    SurfaceHolder.Callback2,
    InputQueue.Callback
{
    val parentApp get() = application as DEngineApp

    lateinit var contentView: NativeView
    // Must be Long to match C++ stuff
    var mWindowId: Long = DEngineApp.INVALID_WINDOW_ID

    // Don't use 0 just to be safe
    private var mAccessibilityIdTracker: Int = 1
    private fun getNextAccessibilityChildId(): Int {
        val temp = mAccessibilityIdTracker
        mAccessibilityIdTracker += 1
        return temp
    }

    private lateinit var mViewConfiguration: ViewConfiguration

    // Don't access directly, the key-value might change. Use wrapper functions
    // There MUST be a 1:1 relationship here
    // This maps the DEngine WidgetID to the virtual child-ID exposed by the AccessibilityNodeInfo
    // If the value exists here, it MUST exist within the accessibility-data
    fun accessibilityWidgetIdFromVirtualAccessChild(virtualAccessChild: Int): Long? {
        val temp = accessibilityData
            .children
            .asSequence()
            .find { it.key.virtualAccessChildId == virtualAccessChild }
        if (temp == null)
            return null
        return temp.key.widgetId
    }

    var accessibilityData = VirtualAccessibilityData(
        focusedId = null,
        accessFocusedId = null,
        children = mapOf(),
        /*
        children = mapOf(
            1L to VirtualAccessibilityElement(
                posX = 0,
                posY = 0,
                width = 400,
                height = 400,
                text = "First"
            ),
            2L to VirtualAccessibilityElement(
                posX = 400,
                posY = 0,
                width = 400,
                height = 400,
                text = "Second"
            ),
            3L to VirtualAccessibilityElement(
                posX = 0,
                posY = 400,
                width = 400,
                height = 400,
                text = "Third"
            ),
        ),
         */
    )

    fun nativeOnNewOrientation(newOrientation: Int) =
        ToNative.onNewOrientation(parentApp.nativeBackendDataPtr, newOrientation)

    fun nativeOnNativeWindowCreated(surface: Surface) {
        ToNative.onNativeWindowCreated(parentApp.nativeBackendDataPtr, surface)
    }
    fun nativeOnNativeWindowDestroyed(surface: Surface) =
        ToNative.onNativeWindowDestroyed(parentApp.nativeBackendDataPtr, surface)

    override fun onCreate(savedState: Bundle?) {
        super.onCreate(savedState)

        parentApp.tryInit(this)

        WindowCompat.enableEdgeToEdge(window)

        mViewConfiguration = ViewConfiguration.get(this)

        contentView = NativeView(this)

        setContentView(contentView)
        contentView.requestLayout()
        contentView.requestFocus()
        contentView.requestApplyInsets()

        window.takeSurface(this)
        // Route this window's raw input (touch, mouse, gamepad, keys) to the native
        // AInputQueue on the game thread instead of the view-hierarchy dispatch. This
        // drives onInputQueueCreated/onInputQueueDestroyed (we implement InputQueue.Callback).
        // IME text still flows through NativeView's InputConnection, independent of this.
        window.takeInputQueue(this)

        val currConfig = Configuration(resources.configuration)
        currOrientation = currConfig.orientation
        currFontScale = currConfig.fontScale
    }


    // Called from the native C++ thread
    fun fromNative_accessibilityUpdate(
        allWidgetId: LongArray,
        allPosX: IntArray,
        allPosY: IntArray,
        allWidth: IntArray,
        allHeight: IntArray,
        allClickable: BooleanArray,
        allTextStart: IntArray,
        allTextCount: IntArray,
        textData: ByteArray)
    {
        runOnUiThread {
            val allLength = listOf(
                allWidgetId.size,
                allPosX.size,
                allPosY.size,
                allWidth.size,
                allHeight.size,
                allClickable.size,
                allTextStart.size,
                allTextCount.size,)
            if (allLength.any { it != allWidgetId.size })
                return@runOnUiThread

            var newAccesChildren = mutableMapOf<WidgetId_VirtualAccessChild_Pair, VirtualAccessibilityElement>()
            // Check if certain WidgetIDs already exist, and reuse them.
            for (index in allWidgetId.indices) {

                val widgetId = allWidgetId[index]

                var virtualChildId = 0
                val temp = accessibilityData.children.keys.find { it.widgetId == widgetId }
                if (temp != null) {
                    virtualChildId = temp.virtualAccessChildId
                } else {
                    virtualChildId = getNextAccessibilityChildId()
                }

                val newChildKey = WidgetId_VirtualAccessChild_Pair(
                    widgetId = widgetId,
                    virtualAccessChildId = virtualChildId,)
                if (newAccesChildren.containsKey(newChildKey)) {
                    Log.e(
                        "DEngine",
                        "Received multiple accessibility element from C++ with widgetId $widgetId")
                    return@runOnUiThread
                }

                val posX = allPosX[index]
                val posY = allPosY[index]
                val width = allWidth[index]
                val height = allHeight[index]
                val clickable = allClickable[index];
                val textStart = allTextStart[index]
                val textCount = allTextCount[index]
                if (textStart < 0 ||
                    textCount < 0 ||
                    textStart + textCount > textData.size)
                {
                    Log.e(
                        "DEngine",
                        "Received accessibility element from C++ with widgetId $widgetId " +
                        "that contained out-of-bounds text slice.")
                    return@runOnUiThread
                }

                val tempString = String(
                    textData,
                    textStart,
                    textCount,
                    Charsets.US_ASCII)

                newAccesChildren.set(
                    newChildKey,
                    VirtualAccessibilityElement(
                        posX = posX,
                        posY = posY,
                        width = width,
                        height = height,
                        clickable = clickable,
                        text = tempString))
            }

            val temp = VirtualAccessibilityData(
                focusedId = null,
                accessFocusedId = null,
                children = newAccesChildren)
            if (accessibilityData == temp) {
                return@runOnUiThread
            }
            accessibilityData = temp



            val accessibilityManager = getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
            if (accessibilityManager.isEnabled) {
                // Send the accessibility event

                val event = AccessibilityEvent()
                event.setSource(contentView)
                event.eventType = AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED
                event.packageName = packageName
                event.className = contentView.javaClass.name
                val success = contentView
                    .parent
                    .requestSendAccessibilityEvent(
                        contentView,
                        event)
                if (!success) {
                    Log.e("DEngine", "Unable to send accessibility event")
                }
            }
        }
    }

    // This is called from the native main game thread.
    @SuppressLint
    fun NativeEvent_HideSoftInput() {
        runOnUiThread(contentView::endInputSession)
    }

    // Surface related callbacks
    override fun surfaceCreated(holder: SurfaceHolder) {
        DEngineApp.verboseEventLog(DEngineApp.EventType.NativeSurfaceCreated)
        nativeOnNativeWindowCreated(holder.surface)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        DEngineApp.verboseEventLog(DEngineApp.EventType.NativeSurfaceChanged)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        DEngineApp.verboseEventLog(DEngineApp.EventType.NativeSurfaceDestroyed)
        nativeOnNativeWindowDestroyed(holder.surface)
    }

    override fun surfaceRedrawNeeded(holder: SurfaceHolder) {
    }
    // Surface related callbacks end

    override fun dispatchGenericMotionEvent(event: MotionEvent?): Boolean {
        if (event == null)
            return false

        for (pointerIndex in 0 until event.pointerCount) {
            val pointerId = event.getPointerId(pointerIndex)
            
            val type = event.getToolType(pointerIndex)
            val source = event.source
            InputDevice.SOURCE_TOUCHSCREEN
            if (type == MotionEvent.TOOL_TYPE_FINGER && (source and InputDevice.SOURCE_TOUCHSCREEN) > 0) {
                // If we get a hover event from tool type finger from touchscreen, it indicates
                // that TalkBack wants us to put accessibility focus on something

                return true
            }
        }

        return false
    }

    // Touch (and all other raw input) is delivered to the native AInputQueue via
    // window.takeInputQueue(this), so we no longer forward it from dispatchTouchEvent.

    /*
    private var prevLocationX = 0
    private var prevLocationY = 0
    private var prevWidth = 0
    private var prevHeight = 0
    override fun onGlobalLayout() {
        if (Build.VERSION.SDK_INT <= 34) {
            val newLocation = IntArray(2)
            contentView.getLocationInWindow(newLocation)

            val newW = contentView.width
            val newH = contentView.height
            val somethingChanged =
                newLocation[0] != prevLocationX ||
                newLocation[1] != prevLocationY ||
                newW != prevWidth ||
                newH != prevHeight

            if (somethingChanged) {
                prevLocationX = newLocation[0]
                prevLocationY = newLocation[1]
                prevWidth = newW
                prevHeight = newH
                nativeOnContentRectChanged(
                    prevLocationX, prevLocationY,
                    prevWidth, prevHeight
                )
            }
        }
    }
     */

    private var currOrientation = 0
    private var currFontScale = 0f
    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (currFontScale != newConfig.fontScale) {
            DEngineApp.verboseEventLog(DEngineApp.EventType.ConfigFontScale)
            currFontScale = newConfig.fontScale
            ToNative.onFontScale(parentApp.nativeBackendDataPtr, mWindowId, currFontScale)
        }

        if (currOrientation != newConfig.orientation) {
            DEngineApp.verboseEventLog(DEngineApp.EventType.ConfigOrientation)
            currOrientation = newConfig.orientation
            ToNative.onNewOrientation(parentApp.nativeBackendDataPtr, currOrientation)
        }
    }

    override fun onInputQueueCreated(queue: InputQueue) {
        ToNative.onInputQueueCreated(
            parentApp.nativeBackendDataPtr,
            this,
            mWindowId,
            queue)
    }

    override fun onInputQueueDestroyed(queue: InputQueue) {
        ToNative.onInputQueueDestroyed(
            parentApp.nativeBackendDataPtr,
            this,
            mWindowId,
            queue)
    }
}